### ISSUE TITLE
[FEATURE] 30 seconds of code - React snippets

### ISSUE THREAD

**Issue by Chalarangelo** [30-seconds/30-seconds-of-react#1]
_From @Chalarangelo on September 10, 2018 20:37_

After some internal discussion and the idea floating around for a bit, I think we should go ahead and brainstorm a bit about adding **React snippets** in the repository, as a separate page, folder etc.

The idea is to provide a basic cheatsheet for React, much like the base repository, except this will be a sub-project (as it is part of the JS ecosystem and we are probably not going to have snippets in the hundreds in this category). I've started compiling a list of popular StackOverflow questions and answers that might prove useful. Add your suggestions below and let's get a decently-sized list ready for October, when [Hacktoberfest](https://hacktoberfest.digitalocean.com/) starts and we can expect a decent flow of PRs.

### My completely incomplete list of ideas

- [x] ~~Collapse #11~~
- [x] ~~Rating #12~~
- [x] ~~Tooltip #22~~
- [x] ~~Modal dialog #17~~
- [x]  ~~`Array.map()` to create a row in a table (or similar) / [SO question](https://stackoverflow.com/a/22877110/1650200) #9 and #26~~
- [x] ~~`Array.map()` to render a list #8~~
- [x] ~~Toggle #20~~
- [x] ~~Autolink (convert link from text to actual `<a>` element with link) #27~~
- [x] ~~Tabs #38~~
- [x] ~~Carousel #23~~
- [x] ~~Drop area file uploader #41~~
- [x] ~~Tree view #43~~
- [x] ~~Timer / Clock / Countdown #33~~
- [x] ~~Slider  8dfdcfd~~
- [x] ~~Accordion c0177c0~~
- [ ] Lazy loader / infinite scroller
- [ ] Overlay / dock
- [ ] Lightbox / image preview
- [ ] Notification / alert
- [ ] Hamburger menu / navigation bar
- [ ] Loader / progress
- [ ] Custom scrollbar
- [ ] Audio / video player
- [ ] Image gallery
- [ ] Calendar
- [ ] Date/time picker
- [ ] Autofocusing input (with additional attributes)
- [ ] Credit card input
- [ ] Color picker
- [ ] Tag input field
- [ ] Rich text editor (support simple things like Bold, Italics, Underline etc.)
- [ ] Image cropper
- [ ] Customizable grid
- [ ] Resizer
- [ ] Combobox

### My completely incomplete list of ideas (old)

- Set focus on input after render using `componentDidMount` / [SO Question](https://stackoverflow.com/a/28890330/1650200)
- Force rerender / [SO Question](https://stackoverflow.com/a/30626072/1650200)
- Pass data on `onClick` event with arrow functions / [SO Question](https://stackoverflow.com/a/29810951/1650200)
- Keying groups of items? / [SO Question](https://stackoverflow.com/q/28329382/1650200)
- Custom version of `shouldComponentUpdate` for component, if we have a good example for that / [SO Question](https://stackoverflow.com/a/24719289/1650200)
- Debounce in React / [SO Question](https://stackoverflow.com/a/28046731/1650200)
- Show/Hide Component / [SO Question](https://stackoverflow.com/a/24534492/1650200)
- Conditional attributes on components / [SO Question](https://stackoverflow.com/a/31164090/1650200)
- Using booleans for `selected <select>` / [SO Question](https://stackoverflow.com/a/21736116/1650200)
- Update array in state / [SO Question](https://stackoverflow.com/a/26254086/1650200)
- Props inside quotes with template literals / [SO Question](https://stackoverflow.com/a/30061326/1650200)
- Pass props to parent / [SO Question](https://stackoverflow.com/a/31756470/1650200)
- How to use `mapDispatchToProps` / [SO Question](https://stackoverflow.com/a/40068198/1650200)
- Properly altering state, using `this.setState`, too many people ask this / [SO Question](https://stackoverflow.com/a/30783011/1650200)
- Call child method from parent / [SO Question](https://stackoverflow.com/a/37950970/1650200)
- Detect click outside component (advanced) / [SO Question](https://stackoverflow.com/a/42234988/1650200)
- Multiple CSS classes for component / [SO Question](https://stackoverflow.com/a/34521945/1650200)
- Conditionally apply classes / [SO Question](https://stackoverflow.com/a/30533260/1650200)
- `setState` with dynamic key name / [SO Question](https://stackoverflow.com/a/29281499/1650200)
- Validate `propTypes` of nested object / [SO Question](https://stackoverflow.com/a/26924696/1650200)
-`setState` with nested objects / [SO Question](https://stackoverflow.com/a/18934259/1650200)
- Component inside component / [SO Question](https://stackoverflow.com/a/31757397/1650200)
- `update` immutability helper / [SO Question](https://stackoverflow.com/a/29537485/1650200)
- Multiple inline styles / [SO Question](https://stackoverflow.com/a/37299564/1650200)
- Remove element from array in state, a bit too basic maybe / [SO Question](https://stackoverflow.com/a/31838774/1650200)
- Add `<script>` via React / [SO Question](https://stackoverflow.com/a/34425083/1650200)
- Form data in React / [SO Question](https://stackoverflow.com/a/23428536/1650200)
- Allow one of multiple types in `propTypes` / [SO Question](https://stackoverflow.com/a/41808543/1650200)
- Add `:hover` styles / [SO Question](https://stackoverflow.com/a/28369384/1650200)
- Dynamic component name, not sure if this works and how / [SO Question](https://stackoverflow.com/a/29876284/1650200)
- Set innerHTML vs dangerouslySetInnerHTML / [SO Question](https://stackoverflow.com/questions/37337289/react-js-set-innerhtml-vs-dangerouslysetinnerhtml)
- Setting default props/ [Devhints](https://devhints.io/react#defaults)
- Returning multiple elements / [Devhints](https://devhints.io/react#new-features)
- `setState` based on previous state / [Comment](https://github.com/30-seconds/30-seconds-of-react/issues/1#issuecomment-426581026)

_Copied from original issue: 30-seconds/30-seconds-of-code#739_

**Comment by Chalarangelo** [comment]
_From @skatcat31 on September 11, 2018 5:31_

You should add Modals as they're extremely helpful to have a dynamic modal that closes on click away from the child

**Comment by Chalarangelo** [comment]
_From @ShawnToubeau on September 18, 2018 13:56_

Hello! Hope you don't mind me adding to this list :) 

- Spread notation / [SO](https://stackoverflow.com/questions/31048953/what-do-these-three-dots-in-react-do)
- Controlled vs uncontrolled input / [SO](https://stackoverflow.com/questions/37427508/react-changing-an-uncontrolled-input)

Noticed a handful of posts asking about React-Router, would that be something you would want to focus on as well?

**Comment by Chalarangelo** [comment]
_From @skatcat31 on September 18, 2018 15:45_

Spread notation is something that's well documented and used in few places, but has definate hlepful places(styles are a big one), and controlled versus uncontrolled input is a huge contention point in React, so it'd be smart to touch base on it. It's something I don't really think about myself since the coding guidelines I implemented enforced controlled at my last job(considered a best practice to guard against injection and cross site)

**Comment by Chalarangelo** [comment]
_From @skatcat31 on September 18, 2018 15:46_

As for organization I think a separate folder or project would be best here. Also to note about Modal is that there are full screen modals based on events, and middle man modals based on parent container. Both would be helpful since I've found more pages struggling with modals.

**Comment by Chalarangelo** [comment]
**Update**: Having tested subdomain support on Github pages, I have to report we *need* to move React snippets into their own repo, as it's the only feasible way to host them under `react.30secondsofcode.org`. Probably a good idea, too, as it will help with separation of concerns. Only issue there is that star count will not be shared, but that's probably alright.

In light of this, I think we should close the issue and go on ahead to create a new repo over the weekend to get started. What do you guys say?

**Comment by Chalarangelo** [comment]
_From @fejes713 on September 27, 2018 15:51_

Better idea to me honestly. 

30code should stay 30code and provide snippets for vanilla JS only. 

30react is a great idea, for the very start, I don't think we need a lot of snippets. Let's keep the focus on quality. If you're up to, we can develop under private repo and then make it public (for the big reveal 💯)

**Comment by Chalarangelo** [comment]
@fejes713 We can't do private repo on org without paying, plus I wanted this to be a good starting point for Hacktoberfest participation, I will make a new repo and see if I can migrate the issue over to it.

**Comment by Chalarangelo** [comment]
_From @fejes713 on September 27, 2018 16:17_

Idea was to make it under user private repo, but as you said Hacktoberfest is here very soon, so we should give people opportunity to make something great!

**Comment by skatcat31** [comment]
Seems like a great place to start. I'm going to be busy the next few weeks but we should at least get a handle seated on how we want to display these and organize content soon

**Comment by meetzaveri** [comment]
@Chalarangelo 

Set innerHTML vs dangerouslySetInnerHTML/[SO Question](https://stackoverflow.com/questions/37337289/react-js-set-innerhtml-vs-dangerouslysetinnerhtml)

**Comment by arjunmahishi** [comment]
Hey! @Chalarangelo I am really looking forward to contributing to this repo. In what format do I submit the snippets?

For example, for `show/hide component`, should the snippet contain the entire class or just the part that is necessary?

**Comment by Chalarangelo** [comment]
We are not set on a snippet template just yet. If you have any template suggestions, you can check #4.

**Comment by SCHKN** [comment]
> _From @ShawnToubeau on September 18, 2018 13:56_
> 
> Hello! Hope you don't mind me adding to this list :)
> 
> * Spread notation / [SO](https://stackoverflow.com/questions/31048953/what-do-these-three-dots-in-react-do)
> * Controlled vs uncontrolled input / [SO](https://stackoverflow.com/questions/37427508/react-changing-an-uncontrolled-input)
> 
> Noticed a handful of posts asking about React-Router, would that be something you would want to focus on as well?

I am not sure that React Router should be added honestly. They have a very great website already displaying a lot of basic and straight-to-the-point examples.  Would be redundant in my opinion.

**Comment by skatcat31** [comment]
> I am not sure that React Router should be added honestly. They have a very great website already displaying a lot of basic and straight-to-the-point examples. Would be redundant in my opinion.

They do, but people often get implementation wrong(especially with redux in the mix). I would suggest that we consider it, but for initial release or scope it can easily be left out(get examples out first and decide if we want to include routers later option)

**Comment by Chalarangelo** [comment]
Yep, I agree with what @skatcat31 said. Let's go for basic snippets first, then see if we want to add router or anything else.

**Comment by nikolalsvk** [comment]
Hey @Chalarangelo and @fejes713, great work on starting this repo and this issue!

I have an idea suggestion for a code snippet, how to `setState` based on the previous state. This is the case where I see most of the people do the wrong thing:

```js
// bad
this.setState({ counter: this.state.counter + 1 })

// good
this.setState(prevState => ({ counter: prevState.counter + 1 }))
```

It's also what ESLint suggests you do if you try the bad practice in your code :)
There's also an explanation in React docs https://reactjs.org/docs/react-component.html#setstate

Let me know what you think and whether it should be added :)

**Comment by Chalarangelo** [comment]
@nikolalsvk That sounds like a useful snippet in general, at least to me.

**Comment by skatcat31** [comment]
It's something I often have to correct junior developers on. Good catch.

**Comment by fejes713** [comment]
@nikolalsvk Definitely a good addition to our collection. Thanks a ton for stopping by to help us! 👍 

@Chalarangelo any plans on when do you want us to start implementing those? What's our current progress with the template?

**Comment by skatcat31** [comment]
Current warning: I have had an extreme amount of interviews this week, and will for the next two on top of moving. I'll help when available but it's looking that that's going to be sparse. A lot of the commenting I'm doing is on scattered access to the internet(has been for awhile... I really just want a consistent schedule again -.-)

**Comment by Chalarangelo** [comment]
@fejes713 Maybe start around October, 15 or something like that? Just to have enough time to decide on a template and figure out everything else. These won't take extremely long anyways. 

If anyone wants to start implementing snippets without a template, feel free to do so, we can reformat them later.

**Comment by fejes713** [comment]
@skatcat31 Good luck on interviews, hopefully, at least one of them will be bingo 😉 

@Chalarangelo Right, I'll try dedicating time around 15th Oct for this 👍

**Comment by SCHKN** [comment]
@Chalarangelo Do you have a place where snippets can be written and reviewed?

**Comment by fejes713** [comment]
@SCHKN I suppose we do not have the exact template for React snippets jet. I would suggest to wait a bit and see what we come up with (less work then - no reformating). 

Once there is a template we will have `../snippets` folder where we will store `snippetName.md` files. In there you can PR snippets that you've written. 

Hopefully this makes sense 😄

TLDR: _by waiting for template to be discussed we won't need to reformat each snippet. Less work = better_

**Comment by nikolalsvk** [comment]
Maybe we don't need to have full components in our snippets 🤔

We can only share relevant idea in the snippet, and make it inputtable (copyable) in any component you like. That's how React docs work and how basically every React library on NPM writes their examples :)

What do y'all think?

**Comment by Chalarangelo** [comment]
I think examples could be made to work in separate codepens or repls either inside the page or opening in new tab, so we can isolate them from the main website code.

**Comment by skatcat31** [comment]
showing relevant parts is definitely a good idea, but it means we'll basically always need a helping post about how our code is organized and what parts are where and such... But either way it should still probably be organized as "things in constructor", "methods", etc. and have some sort of common order to snippets.

**Comment by nikolalsvk** [comment]
"Help" post would be awesome to have, to explain the structure of the cards :)

Plus, it's not a big deal to implement a help box at each page of the future website with nowdays compilers of Markdown files.

**Comment by Chalarangelo** [comment]
**Note:** Reviewing the suggested snippets, I can see that many of those suggestions are not exactly things that we want to provide as components, but rather concepts we want to show off in components that we want to implement. Therefore, we still need a few suggestions for snippets, so we can find a way to integrate those concepts into new components.

**Comment by meetzaveri** [comment]
Yes I agree @Chalarangelo . Some of those are/could be part of a component. Some of them might be useful as a component. 

The concepts can also be **mulit-categorized** in future as this project expands i.e. fall in one or more than one domain. Examples like those additional domains would be - redux, react-router, formik, PropTypes and much more. 

So we need those domains to interact with this project which provides information about only **"React.js"**(framework's perspective)

**Comment by Chalarangelo** [comment]
Redux, React-Redux and React-Router should ideally be part of the same repo as React. Same applies to PropTypes. The rest of the tools are probably out of scope for this particular project and should be explored later down the line. However, currently, we are focusing on simple React snippets to get things right first, before expanding and combining with other tools.

**Comment by skatcat31** [comment]
So we should decide on a final snippet structure so we can start applying it to the currently suggested snippets. I've been reviewing a few but my time has been limited due to the interview process.

**Comment by jonniespratley** [comment]
What’s the status on this? I can contribute and need to know the snippet format.

**Comment by Chalarangelo** [comment]
@jonniespratley the format is the one described on the snippet template file (or you can take a look at existing snippets, as well).

**Comment by Ankitb09** [comment]
Hey @Chalarangelo @skatcat31 I wanted to contribute in this. Can you please tell me how to make my contribution in this awesome library?

**Comment by arjunmahishi** [comment]
> Hey @Chalarangelo @skatcat31 I wanted to contribute in this. Can you please tell me how to make my contribution in this awesome library?

You can have a look at the Issues. You can mainly contribute snippets or there are other "Chores"

**Comment by skatcat31** [comment]
I'm going to add another chore for `Contributing Guidelines`. Currently I'm on boarding at work, so maybe in a few weeks I can circle back around to this after I get unemployment and insurance figured.

**Comment by fejes713** [comment]
@meetzaveri that's a good point. Totally agree.

I was reading a few articles about clean code and code reusability. After semi-intensive research, I still believe that we should use prop types and default props. **Yes, it will make the card longer** but user shouldn't try to guess the props or look into the code itself - bad habits. 

Remeber 30s of code? You just copy paste the snippet and it works. If we don't follow this approach with 30s of react somebody will copy it without any props and get error. Think about development experience and robustness of the snippet. The goal is to build reusable components, not the halfway done snippet just to say we did something.

Think about it few times; let's go for quality not quantity. @skatcat31 @Chalarangelo @meetzaveri

**Comment by Chalarangelo** [comment]
@fejes713 the general issue with default props is that some things cannot be set to sensible defaults (e.g. content of components). I'm not against this, I'm just trying to figure out what the best course of action is.

**Comment by skatcat31** [comment]
@fejes713 That's why I say any defaults should be assigned in the constructor. Default props, while useful, is still awkward in that it's not actually on the class, and makes copy pasta harder because you have to remember to look for other things too. I'm a big fan of JS classes keeping instantiations to the constructor(hence my gripe with field proposal)

**Comment by theodorusyoga** [comment]
Hi @Chalarangelo , first-timer here. This is an awesome repo and I am looking to contribute to this repo! I'm interested in one of these: lazy loading component, notifications, or hamburger menu. I'm already read the contributing guidelines mentioned above. Can you tell me which one you recommend more and the details that you are trying to achieve? Thanks!

**Comment by fejes713** [comment]
Hey @theodorusyoga, thanks for willing to help us! We're glad that you like the project 😄 

We're trying to make a collection of useful components that you can integrate to your own project and use for learning purposes. 

Pick whatever component you feel most comfortable writing and I'll be happy to help out and review your PR.

Let me know if you need any help.

**Comment by Chalarangelo** [comment]
This issue is a bit of a mess right now, I'm closing it but we can still add snippets suggested here in new PRs.
